substrate/staking/pallet/src/lib.rs

```diff
@@ -0,0 +1,219 @@
+      Staked::<T>::mutate(account, |staked| {
+        let available = *staked - Self::allocated(account);
+        // Check this invariant in the name of safety
+        if available < amount {
+          Err(Error::<T>::BondUnavailable)?;
+        }
+        *staked -= amount;
+        Ok::<_, DispatchError>(())
+      })
+    }
+
+    fn allocate_internal(account: &T::AccountId, amount: u64) -> Result<(), Error<T>> {
+      Allocated::<T>::try_mutate(account, |allocated| {
+        let available = Self::staked(account) - *allocated;
+        if available < amount {
+          Err(Error::<T>::BondUnavailable)?;
+        }
+        *allocated += amount;
+        Ok(())
+      })
+    }
+
+    fn deallocate_internal(account: &T::AccountId, amount: u64) -> Result<(), Error<T>> {
+      Allocated::<T>::try_mutate(account, |allocated| {
+        if *allocated < amount {
+          Err(Error::<T>::InSufficientAllocation)?;
+        }
+        *allocated -= amount;
+        Ok(())
+      })
+    }
+  }
+
+  #[pallet::call]
+  impl<T: Config> Pallet<T> {
+    /// Stake funds from this account.
+    #[pallet::call_index(0)]
+    #[pallet::weight((0, DispatchClass::Operational))] // TODO
+    pub fn stake(origin: OriginFor<T>, #[pallet::compact] amount: u64) -> DispatchResult {
+      let signer = ensure_signed(origin)?;
+      // Serai accounts are solely public keys. Accordingly, there's no harm to letting accounts
+      // die. They'll simply be re-instantiated later
+      // AllowDeath accordingly to not add additional requirements (and therefore annoyances)
+      T::Currency::transfer(&signer, &Self::account(), amount, ExistenceRequirement::AllowDeath)?;
+      Self::add_stake(&signer, amount);
+      Ok(())
+    }
+
+    /// Unstake funds from this account. Only unallocated funds may be
+    /// unstaked.
+    #[pallet::call_index(1)]
+    #[pallet::weight((0, DispatchClass::Operational))] // TODO
+    pub fn unstake(origin: OriginFor<T>, #[pallet::compact] amount: u64) -> DispatchResult {
+      let signer = ensure_signed(origin)?;
+      Self::remove_stake(&signer, amount)?;
+      // This should never be out of funds as there should always be stakers. Accordingly...
+      // TODO: What if this fails for some reason but we removed the stake above?
+      T::Currency::transfer(&Self::account(), &signer, amount, ExistenceRequirement::KeepAlive)?;
+      Ok(())
+    }
+
+    /// Allocate `amount` to a given validator set.
+    #[pallet::call_index(2)]
+    #[pallet::weight((0, DispatchClass::Operational))] // TODO
+    pub fn allocate(
+      origin: OriginFor<T>,
+      network: NetworkId,
+      #[pallet::compact] amount: u64,
+    ) -> DispatchResult {
+      let account = ensure_signed(origin)?;
+
+      // add to amount bonded
+      Self::allocate_internal(&account, amount)?;
+
+      // add to participants list for the network
+      let result = VsPallet::<T>::add_participant(account, Amount(amount), network);
```

Comment:
`handle_allocation`? since then inside of the function, we either add or increase.

Reply:
Actually no handle is very general, it might also mean the decrease. So I guess best is `increase_allocation`, since also can assume that in the case of add, validator had `0` allocation an now has more than `0`.